sqlgen/matcher.go

```diff
@@ -0,0 +1,118 @@
+package sqlgen
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/samsarahq/thunder/internal"
+)
+
+// extractColumns extracts the columns used in a filter.
+func extractColumns(filter Filter) []string {
+	columns := make([]string, 0, len(filter))
+	for column := range filter {
+		columns = append(columns, column)
+	}
+	sort.Strings(columns)
+	return columns
+}
+
+// extractColumnValues extracts the values of a set of columns in a filter.
+func extractColumnValues(m map[string]interface{}, columns []string) []interface{} {
+	values := make([]interface{}, 0, len(columns))
+	for _, column := range columns {
+		values = append(values, m[column])
+	}
+	return values
+}
+
+// columnsKey builds a string key for a set of columns useable in a map.
+func columnsKey(columns []string) string {
+	return strings.Join(columns, ";")
+}
+
+// A matcherGroup holds a set of tuples for a given set of columns in a
+// matcher.
+type matcherGroup struct {
+	columns []string
+	tuples  map[interface{}]map[interface{}]struct{}
+}
+
+// A matcher efficiently matches table rows against sets of queries. For every
+// query shape (set of columns that a query matches), the matcher holds a
+// lookup table from column values tuples to queries.
+type matcher struct {
+	groups map[string]*matcherGroup
+}
+
+// newMatcher creates a new matcher.
+func newMatcher() *matcher {
```

Comment:
I'd like to use this matcher also in livesql's binlog change detection, but that's not going to happen right now.